### PR TITLE
Addition of new script update_workflow.py

### PR DESCRIPTION
**Pull request by svenwillger**
To update a workflow I listed all files of the 'template_snakemake' repo except for the files "/workflow/rules/00_modules.smk" and "/workflow/rules/99_aggregates.smk" because they are project specific and have been modified specifically for the project.

I added the parser "branch" that allows to provide the desired branch or tag from which the workflow should be updated from and the necessary changes throughout the script.

I also included a check if the chosen branch/tag exists (requires to import the package 'requests') which I also included in the 'cubi-tools.yaml'. Also made some changes to the 'report_script_version' function because it wasn't working.

Open issue. If using the --external function, only the root folder is getting created in the "cubi" folder, but the subdirectories are not created causing the problem that files from subdirectories are not getting downloaded. If those subdirectories are created manually, the download of the files works properly.

### CODE REVIEW

**Comment by ptrebert** [comment]
I will add a few specific remarks in the code files, but here are a few general things:

1) You raised the point yourself that fetching everything from the remote (github) hits the API rate limit even though we are not yet doing many things. Hence, the implementation should pull the template repository with something like `git pull --all` and then update from the local copy to avoid rate limiting.
2) The code is not readily designed to run in batches, i.e. the call to update the workflow files should be encapsulated in a function that can be called on a workflow directory (plus some parameters if necessary).
3) repeated calls to standard library functions such as `subprocess.run` should always be encapsulated in a separate function (see actually how this is done (e.g., see how this is done in the Snakemake template for calls to `rsync` or for writing log messages in the `commons::pyutils.py` module).
4) The code has not been reformatted with `black`; the changes are minor, though
5) None of the functions actually has a docstring explaining what the function should do - it's really difficult to follow the logic of the code

File: cubi-tools/update_workflow.py [hunk 1]
```diff
@@ -14,7 +15,16 @@ def main():
     print(f"Project directory set as: {project_dir}")
     ref_repo_clone = args.ref_repo_clone
```
Comment (ptrebert) [inline cubi-tools/update_workflow.py:16]: my IDE tells me that this variable is never used - what does it do?

File: cubi-tools/update_workflow.py [hunk 2]
```diff
@@ -14,7 +15,16 @@ def main():
...
     ref_repo_curl = args.ref_repo_curl
-    ref_repo_wget = args.ref_repo_wget
+    branch = args.branch
+    ref_repo_wget = args.ref_repo_wget + branch + "/"
+
+    #detect if branch exists
+    response = requests.get(f"{ref_repo_wget}README.md")
+    if response.status_code != 200:
+        raise Exception("That is not a valid branch or tag")
+    else:
```
Comment (ptrebert) [inline cubi-tools/update_workflow.py:25]: totally unnecessary else here

File: cubi-tools/update_workflow.py [hunk 3]
```diff
@@ -33,6 +43,17 @@ def main():
         "CITATION.md",
         "LICENSE",
         ".editorconfig",
+        "init.py",
```
Comment (ptrebert) [inline cubi-tools/update_workflow.py:46]: here is a logical problem - this script should update the workflow template, which includes the metadata, yes, but the update of the metadata is realized in a different script. What this results in is one big pile of duplicated code. So focus here first only on the template update and leave the metadata update in the domain of the other script.

File: cubi-tools/update_workflow.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import subprocess as sp
 import argparse as argp
 import toml
+import requests
```
Comment (ptrebert) [inline cubi-tools/update_workflow.py:8]: you introduced an additional dependency and I don't see why that is necessary, you only use it to check if the repo exists, which can simply be done by trying to access it and assuming it's invalid if that fails.

File: envs/conda/cubi_tools.yaml [hunk 5]
```diff
@@ -10,3 +10,4 @@ dependencies:
   - pylint=2.14.5
   - isort=5.10.1
   - black=22.6.0
+  - requests=2.25.1
```
Comment (ptrebert) [inline envs/conda/cubi_tools.yaml:13]: see my comment above - only add packages/libs if they provide vital functionality

File: cubi-tools/update_workflow.py (hunk 1 above)
Comment (svenwillger) [reply cubi-tools/update_workflow.py]: The whole section has been modified.

File: cubi-tools/update_workflow.py (hunk 2 above)
Comment (svenwillger) [reply cubi-tools/update_workflow.py]: The whole check if the repo/branch exist has been replaced

File: cubi-tools/update_workflow.py (hunk 3 above)
Comment (svenwillger) [reply cubi-tools/update_workflow.py]: A switch has been implemented so the user can select to also update the metadata; however, the default is set to False

File: cubi-tools/update_workflow.py (hunk 4 above)
Comment (svenwillger) [reply cubi-tools/update_workflow.py]: Has been removed

File: envs/conda/cubi_tools.yaml (hunk 5 above)
Comment (svenwillger) [reply envs/conda/cubi_tools.yaml:13]: Has been removed

**Comment by svenwillger** [comment]
The 'update_workflow.py' has been modified significantly and addressed all raised issues.
Docstrings has been added to all functions, the script has been formatted with `black`.
Screen messages are clearer regarding what is done and also the outcome of those actions is printed.

**Pull request closed without merging**